src/os/unix/wayland.rs

```diff
@@ -0,0 +1,1117 @@
+
+        // Acknowledge the last configure event
+        if let Some(serial) = (*self.xdg_config.borrow_mut()).take() {
+            self.xdg_surface.ack_configure(serial);
+        }
+
+        self.surface.attach(Some(buf), 0, 0);
+        self.surface
+            .damage(0, 0, i32::max_value(), i32::max_value());
+        self.surface.commit();
+    }
+
+    fn get_toplevel_info(&self) -> (ToplevelResolution, ToplevelClosed) {
+        let configure = Rc::new(RefCell::new(None));
+        let close = Rc::new(RefCell::new(false));
+
+        {
+            let configure = configure.clone();
+            let close = close.clone();
+
+            self.toplevel.quick_assign(move |_, event, _| {
+                use wayland_protocols::xdg_shell::client::xdg_toplevel::Event;
+
+                if let Event::Configure { width, height, .. } = event {
+                    *configure.borrow_mut() = Some((width, height));
+                } else if let Event::Close = event {
+                    *close.borrow_mut() = true;
+                }
+            });
+        }
+
+        (configure, close)
+    }
+}
+
+struct WaylandInput {
+    kb_events: mpsc::Receiver<wl_keyboard::Event>,
+    pt_events: mpsc::Receiver<wl_pointer::Event>,
+    _keyboard: Main<WlKeyboard>,
+    pointer: Main<WlPointer>,
+}
+
+impl WaylandInput {
+    fn new(seat: &Main<WlSeat>) -> Self {
+        let (keyboard, pointer) = (seat.get_keyboard(), seat.get_pointer());
+
+        let (kb_sender, kb_receiver) = mpsc::sync_channel(1024);
+
+        keyboard.quick_assign(move |_, event, _| {
+            kb_sender.send(event).unwrap();
+        });
+
+        let (pt_sender, pt_receiver) = mpsc::sync_channel(1024);
+
+        pointer.quick_assign(move |_, event, _| {
+            pt_sender.send(event).unwrap();
+        });
+
+        Self {
+            kb_events: kb_receiver,
+            pt_events: pt_receiver,
+            _keyboard: keyboard,
+            pointer,
+        }
+    }
+
+    fn get_pointer(&self) -> &Main<WlPointer> {
+        &self.pointer
+    }
+
+    fn iter_keyboard_events(&self) -> mpsc::TryIter<wl_keyboard::Event> {
+        self.kb_events.try_iter()
+    }
+
+    fn iter_pointer_events(&self) -> mpsc::TryIter<wl_pointer::Event> {
+        self.pt_events.try_iter()
+    }
+}
+
+pub struct Window {
+    display: DisplayInfo,
+
+    width: i32,
+    height: i32,
+
+    scale: i32,
+    bg_color: u32,
+    scale_mode: ScaleMode,
+
+    mouse_x: f32,
+    mouse_y: f32,
+    scroll_x: f32,
+    scroll_y: f32,
+    buttons: [u8; 3],
+    prev_cursor: CursorStyle,
+
+    should_close: bool,
+    active: bool,
+
+    key_handler: KeyHandler,
+    // Option because MaybeUninit's get_ref() is nightly-only
+    keymap: Option<xkb::keymap::Keymap>,
+    update_rate: UpdateRate,
+    menu_counter: MenuHandle,
+    menus: Vec<UnixMenu>,
+    input: WaylandInput,
+    resizable: bool,
+    // Temporary buffer
+    buffer: Vec<u32>,
+    // Configure, close
```

Comment:
Just to confirm my suspicions: `ToplevelClosed` means whether or not the window is closed ("minimized"), correct?

```suggestion
    // Resolution, closed
```

Reply:
Yes whether or not the client wants to close the window.